web/src/components/TopicModal.jsx

```diff
@@ -211,53 +211,118 @@ export function TopicModal(props) {
+        misp_tags:
+          presetMispTag?.length > 0
+            ? presetMispTag.split(",").map((mispTag) => mispTag.trim())
+            : [],
+      };
+
+      return JSON.stringify(data) !== JSON.stringify(presetData) && src.created_by === user.user_id;
+    }
+
+    function getNeedUpdateAction(_actions, _presetActionIds) {
+      const currentActionIds = _actions.map((action) => action.action_id);
+      const presetActionIdsArray = [..._presetActionIds];
+
+      if (!arrayComparison(currentActionIds, presetActionIdsArray)) return true;
+
+      for (let i = 0; i < _actions.length; i++) {
+        if (_actions[i].recommended !== presetActions[i].recommended) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+    async function updateTopicPromise() {
+      return updateTopic(topicId, data).catch((error) => operationError(error));
+    }
+
+    async function updateActionPromise() {
+      let promiseArray = [];
+      for (let action of actions) {
+        const actionRequest = {
+          ...action,
+          topic_id: topicId,
+        };
+        if (action.action_id === null) {
+          promiseArray.push(createAction(actionRequest).catch((error) => operationError(error)));
+        } else if (presetActionIds.has(action.action_id)) {
+          presetActionIds.delete(action.action_id);
+          promiseArray.push(
+            updateAction(action.action_id, actionRequest).catch((error) => operationError(error)),
+          );
+        }
+      }
+
+      // delete actions that are not related to topic
+      for (let actionId of presetActionIds) {
+        promiseArray.push(deleteAction(actionId).catch((error) => operationError(error)));
+      }
+      if (src.created_by !== user.user_id) {
+        enqueueSnackbar(
+          "Only actions have been changed, not topics. You can't update topic, because you are not topic creator.",
+          { variant: "warning" },
+        );
+      }
+      return Promise.all(promiseArray);
+    }
+
+    const needUpdateAction = getNeedUpdateAction(actions, presetActionIds);
+    const needUpdateTopic = getNeedUpdateTopic();
+
+    if (needUpdateAction) {
+      if (needUpdateTopic) {
+        Promise.all([updateActionPromise(), updateTopicPromise()]).then(() => {
+          reloadTopicAfterAPI();
+          enqueueSnackbar("Update topic and action succeeded", { variant: "success" });
+          onSetOpen(false);
+        });
+      } else {
+        updateActionPromise().then(() => {
+          reloadTopicAfterAPI();
+          enqueueSnackbar("Update action succeeded", { variant: "success" });
+          onSetOpen(false);
+        });
+      }
+    } else {
+      if (needUpdateTopic) {
+        updateTopicPromise().then(() => {
+          reloadTopicAfterAPI();
+          enqueueSnackbar("Update topic succeeded", { variant: "success" });
+          onSetOpen(false);
+        });
+      }
+    }
```

Comment:
Topicとactionの同時更新を分類する必要はありますか？シンプルに以下のように書くのではダメですかね？
```suggestion
    if (needUpdateAction) {
      updateActionPromise();
      enqueueSnackbar("Update action succeeded", { variant: "success" });
   }
   
   if (needUpdateTopic) {
          updateTopicPromise
          reloadTopicAfterAPI();
          enqueueSnackbar("Update action succeeded", { variant: "success" });
   }
   onSetOpen(false);
```

promiseを返して無理やりここで Promise.All をやる理由があるのかがあまり読み取れませんでした